images/bridge2/assistant/assistant.go

```diff
@@ -24,52 +34,169 @@ type AssistantTextEvent struct {
+	}
+
+	if a.SessionStorage != nil {
+		sync := newFSSync(sess, a.SessionStorage.SessionStoragePath(sess, "assistant"), 500*time.Millisecond)
+		sess.Listen(sync)
+	}
 }
 
 func (a *Agent) HandleEvent(annot tracks.Event) {
 	if annot.Type != "transcription" {
 		return
 	}
+	handleTranscription(a.DefaultAssistants, annot)
+}
+
+type sessionAgent struct {
+	mu         sync.RWMutex
+	NewClient  func(name, prompt string) Client
+	assistants map[string]eventClient
+}
+
+type eventClient struct {
+	Client Client
+	Event  tracks.EventMeta
+}
+
+func (a *sessionAgent) HandleEvent(annot tracks.Event) {
+	switch annot.Type {
+	case "assistant-set-prompt":
+		a.handleSetPrompt(annot)
+	case "transcription":
+		handleTranscription(a.getAssistants(), annot)
+	}
+}
+
+func eventBefore(a, b tracks.EventMeta) bool {
+	if a.Start < b.Start {
+		return true
+	}
+	if a.Start > b.Start {
+		return false
+	}
+	return a.ID.Compare(b.ID) < 0
+}
+
+func (a *sessionAgent) handleSetPrompt(annot tracks.Event) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	in := annot.Data.(*AssistantPromptEvent)
+	prev, hasPrev := a.assistants[in.Name]
+	if hasPrev && eventBefore(annot.EventMeta, prev.Event) {
+		log.Printf("assistant: not updating %q, existing client from %s newer than %s", in.Name, prev.Event.Start, annot.Start)
+		return
+	}
+	record := eventClient{
+		Event: annot.EventMeta,
+	}
+	if in.SystemMessage != "" {
+		record.Client = a.NewClient(in.Name, in.SystemMessage)
+	}
+	if hasPrev {
+		log.Printf("assistant: replacing %q client from %s with newer %s", in.Name, prev.Event.Start, annot.Start)
+	} else {
+		log.Printf("assistant: adding %q client at %s", in.Name, annot.Start)
+	}
+	a.assistants[in.Name] = record
+}
+
+func (a *sessionAgent) getAssistants() []Client {
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+	out := make([]Client, 0, len(a.assistants))
+	for _, client := range a.assistants {
+		if client.Client != nil {
+			out = append(out, client.Client)
+		}
+	}
+	return out
+}
+
+func handleTranscription(clients []Client, annot tracks.Event) {
 	in := annot.Data.(*transcribe.Transcription)
 
 	if len(in.Segments) == 0 || len(in.Segments[0].Words) == 0 {
 		log.Printf("assistant: transcription had no segments or words")
 		return
 	}
 	text := strings.TrimSpace(in.Text())
-	names := a.matchAssistants(text)
+	names := matchAssistants(clients, text)
 	log.Printf("assistant: matched %v for: %s", names, text)
-	for _, name := range names {
-		go a.respond(annot, name, text)
+	for name, client := range names {
+		go respond(client, annot, name, text)
 	}
 }
 
-func (a *Agent) matchAssistants(text string) []string {
-	var matched []string
+func matchAssistants(clients []Client, text string) map[string]Client {
+	matched := make(map[string]Client)
 	text = strings.ToLower(text)
-	for name := range a.Assistants {
-		if strings.Contains(text, name) {
-			matched = append(matched, name)
+	for _, client := range clients {
+		if strings.Contains(text, client.AssistantName()) {
```

Comment:
Ok, I have one commit that does this, though as I was looking into storing prompts, I combined it so that there's just one call to `Complete` which can return an `ErrNoMatch` for inputs that it doesn't support. As I noted in the commit message, if we want to keep it as separate steps, the "match" step could return what the prompt would be, or no prompt if it's not supported. Then we pass the prompts back in for completion.

Reply:
one step together works for me